packages/ckeditor5-source-editing/src/sourceediting.js

```diff
@@ -0,0 +1,320 @@
+ *
+ * It provides the possibility to view and edit the source of the document.
+ *
+ * For a detailed overview, check the {@glink features/source-editing source editing feature documentation} and the
+ * {@glink api/source-editing package page}.
+ *
+ * @extends module:core/plugin~Plugin
+ */
+export default class SourceEditing extends Plugin {
+	/**
+	 * @inheritDoc
+	 */
+	static get pluginName() {
+		return 'SourceEditing';
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	static get requires() {
+		return [ PendingActions ];
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	constructor( editor ) {
+		super( editor );
+
+		/**
+		 * Flag indicating whether the document source mode is active.
+		 *
+		 * @observable
+		 * @member {Boolean}
+		 */
+		this.set( 'isSourceEditingMode', false );
+
+		/**
+		 * The element replacer instance used to replace the editing roots with the wrapper elements containing the document source.
+		 *
+		 * @private
+		 * @member {module:utils/elementreplacer~ElementReplacer}
+		 */
+		this._elementReplacer = new ElementReplacer();
+
+		/**
+		 * Maps all root names to wrapper elements containing the document source.
+		 *
+		 * @private
+		 * @member {Map.<String,HTMLElement>}
+		 */
+		this._replacedRoots = new Map();
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	init() {
+		const editor = this.editor;
+		const t = editor.t;
+
+		editor.ui.componentFactory.add( 'sourceEditing', locale => {
+			const buttonView = new ButtonView( locale );
+
+			buttonView.set( {
+				label: t( 'Edit source' ),
+				// TODO: use icon
+				// icon: sourceEditingIcon,
+				tooltip: true,
+				withText: true,
+				class: 'ck-source-editing'
+			} );
+
+			buttonView.bind( 'isOn' ).to( this, 'isSourceEditingMode' );
+
+			// The button should be disabled if one of the following condition is met.
+			buttonView.bind( 'isEnabled' ).to(
+				this, 'isEnabled',
+				editor, 'isReadOnly',
+				editor.plugins.get( PendingActions ), 'hasAny',
+				( isEnabled, isEditorReadOnly, hasAnyPendingActions ) => {
+					// (1) The plugin is disabled itself.
+					if ( !isEnabled ) {
+						return false;
+					}
+
+					// (2) The editor is in a read-only mode.
+					if ( isEditorReadOnly ) {
+						return false;
+					}
+
+					// (3) Any pending action is scheduled. It may change the model, so modifying the document source should be prevented
+					// until the model is finally set.
+					if ( hasAnyPendingActions ) {
+						return false;
+					}
+
+					return true;
+				}
+			);
+
+			this.listenTo( buttonView, 'execute', () => {
+				this.isSourceEditingMode = !this.isSourceEditingMode;
+
+				/**
+				 * Fired whenever the source editing mode is toggled.
+				 *
+				 * @event sourceEditing
+				 * @param {Object} data Additional information about the event.
+				 * @param {Boolean} data.isSourceEditingMode Flag indicating whether the document source mode is active.
+				 */
+				this.fire( 'sourceEditing', { isSourceEditingMode: this.isSourceEditingMode } );
```

Comment:
As above.

Reply:
Custom `sourceEditing` is no longer fired.